File: Core/src/util/ThreadUtil.java
package util;

import java.lang.Thread.State;
import java.util.Iterator;
import java.util.Map;

public class ThreadUtil
{
	public static void dumpStacks()
	{
		Debug.appendWithoutDate("");
		Debug.appendBanner("STACKS DUMP");
		
		Map<Thread, StackTraceElement[]> threads = Thread.getAllStackTraces();
		Iterator<Thread> it = threads.keySet().iterator();
		for (; it.hasNext(); )
		{
			Thread thread = it.next();
			StackTraceElement[] stack = thread.getStackTrace();
			State state = thread.getState();
			if (stack.length > 0)
			{
				Debug.append("---- THREAD " + thread.getName() + "  (" + state + ") ----");
				for (StackTraceElement element : stack)
				{
					Debug.appendWithoutDate("" + element);
				}
			}
		}
		
		Debug.appendWithoutDate("");
	}
	
	public static void sleepWithCatch(long millis)
	{
		try
		{
			Thread.sleep(millis);
		}
		catch (InterruptedException ie)
		{
			Debug.append("Caught " + ie + " sleeping for " + millis + " millis");
		}
	}
}

File: Core/src/burlton/core/code/util/ThreadUtil.kt
package burlton.core.code.util

import javax.swing.SwingUtilities

fun dumpThreadStacks()
{
    Debug.appendWithoutDate("")
    Debug.appendBanner("STACKS DUMP")

    val threads = Thread.getAllStackTraces()
    val it = threads.keys.iterator()
    while (it.hasNext())
    {
        val thread = it.next()
        val stack = thread.stackTrace
        val state = thread.state
        if (stack.isNotEmpty())
        {
            Debug.append("---- THREAD " + thread.name + "  (" + state + ") ----")
            for (element in stack)
            {
                Debug.appendWithoutDate("" + element)
            }
        }
    }

    Debug.appendWithoutDate("")
}

fun runOnEventThread(r: (() -> Unit))
{
    if (SwingUtilities.isEventDispatchThread())
    {
        r.invoke()
    }
    else
    {
        SwingUtilities.invokeLater(r)
    }
}